File: 200-number-of-islands/number-of-islands.java
class Solution {
    public int numIslands(char[][] grid) {
        int count = 0;
        boolean[][] visited = new boolean[grid.length][grid[0].length];

        for(int row = 0; row < grid.length; row++) {
            for(int col = 0; col < grid[row].length; col++) {
                if(grid[row][col] == '1') {
                    count++;
                    traverseBFS(grid, row, col);
                }
            }
        }

        return count;
    }

    private void traverseBFS(char[][] grid, int row, int col) {
        int[][] directions = {{-1, 0}, {0, 1}, {1, 0}, {0, -1}};

        LinkedList<Integer> queue = new LinkedList<>();
        queue.add(row);
        queue.add(col);

        while(!queue.isEmpty()) {
            row = queue.removeFirst();
            col = queue.removeFirst();
            grid[row][col] = '0';

            for(int i = 0; i < directions.length; i++) {
                int newRow = row + directions[i][0];
                int newCol = col + directions[i][1];

                if(newRow < 0 || newRow >= grid.length || newCol < 0 || newCol >= grid[0].length) {
                    continue;
                }

                if(grid[newRow][newCol] != '0') {
                    queue.add(newRow);
                    queue.add(newCol);
                    grid[newRow][newCol] = '0';
                }
            }
        }
    }
}

File: 200-number-of-islands/number-of-islands.py
class Solution:
    def dfs(self, i, j, visited, grid):
        if i < 0 or i >= len(visited): return
        if j < 0 or j >= len(visited[0]): return
        if grid[i][j] == "0": return
        if visited[i][j] == 1: return 

        visited[i][j] = 1

        self.dfs(i+1,j,visited,grid)
        self.dfs(i-1,j,visited,grid)
        self.dfs(i,j-1,visited,grid)
        self.dfs(i,j+1,visited,grid)

    def numIslands(self, grid: List[List[str]]) -> int:
        result = 0
        visited = [[0 for _ in range(len(grid[0]))] for __ in range(len(grid))]
        for i in range(len(grid)):
            for j in range(len(grid[0])):
                if visited[i][j] == 0 and grid[i][j] == "1":
                    self.dfs(i,j,visited, grid)
                    result += 1

        return result 